html/user/openssl_sign_action.php

```diff
@@ -0,0 +1,148 @@
+
+	// How to generate required keys in /project/keys/ folder:
+	// openssl genpkey -algorithm RSA -out external_system_key_private.pem -pkeyopt rsa_keygen_bits:2048
+	// openssl rsa -pubout -in external_system_key_private.pem -out external_system_key_public.pem
+
+	// FAO project admins: Change the following 2 lines to the location of the keys you have personally generated.
+	// If the following keys do not exist, then the users will face (handled) error messages.
+	$private_key_path = "file:///root/project/keys/external_system_key_private.pem";
+	$public_key_path = "file:///root/project/keys/external_system_key_public.pem";
+
+	// Check that the private key file exists where specified. If not, redirect to error page.
+	if (!file_exists($private_key_path)) {
+		error_page(tra("The private key does not exist. Proof of UserID currently not possible."));
+	}
+
+	// Check that the public key file exists where specified. If not, redirect to error page.
+	if (!file_exists($public_key_path)) {
+		error_page(tra("The public key does not exist. Proof of UserID currently not possible."));
+	}
+
+	// Fetch private key from file
+	$private_key_pem = openssl_pkey_get_private($private_key_path);
+
+	// Compute signature using the SHA512 signing algorithm
+	$openssl_signed_message = openssl_sign($message_data, $signature, $private_key_pem, OPENSSL_ALGO_SHA512);
+
+	// Free the private key from memory
+	openssl_free_key($private_key_pem);
+	
+	// Check that the signature exists
+	if ($openssl_signed_message == 1) {
+		// The signature was successfully created.
+
+		// fetch public key into memory
+		$pubkeyid = openssl_pkey_get_public($public_key_path);
+
+		// Base64 encode the generated signature to enable safe output to text file.
+		$base64_sig = base64_encode($signature);
+
+		// Decode base64 sig for use in sig_verification
+		$decoded_sig = base64_decode($base64_sig);
+
+		// Verify that the generated signature against the original data, using the public key.
+		$sig_verification = openssl_verify($message_data, $decoded_sig, $pubkeyid, OPENSSL_ALGO_SHA512);
+		
+		// Free the public key from memory
+		openssl_free_key($pubkeyid);
+
+		// Check if signature was successfully validated
+		if ($sig_verification == 1) {
+			// The generated signature has been successfully verified using the public key.
+
+			// Generate an signature xml file auto-download prompt
+			header("Content-Disposition: attachment; filename=signature.xml");
+			header("Content-type: text/xml");
+
+			// Start outputting xml to the auto-download xml file
+			echo "<boinc_user_id_verification>";
+			global $master_url; // Define global master_url variable for use in xml output.
+			echo "\n <master_url>$master_url</master_url>";
+			echo "\n <userid>$userid</userid>";
+			echo "\n <user_data>$user_data</user_data>";
+			//echo "<cpid>$cpid</cpid>"
+			//echo "<timestamp>time()</timestamp>"
+			echo "\n <msg>$message_data</msg>";
```

Comment:
Either remove user_data or msg elements. This is duplicate data. Or it does not matter? Idk.

Reply:
The msg element is hashed, so you can't see the original user_data contents. Fair enough the external service (like the Gridcoin wallet) should know what the intended/original user_data was, but perhaps it'd be easier to import the XML and read the user_data field from there for quick reference/double-checking?

Overall, the performance improvement by removing this data is negligible, the XML download prompt is almost instant after you click the 'submit' button.

Reply:
The msg element contains contents of both user_id and user_data elements. So it is kind of redundant there. You removed the digest/hashing step there, or is it?

Reply:
Ah, sorry I see what you mean I didn't have the XML downloaded file to hand earlier, yeah only the 'signature' has the signed message, so the 'userid' and 'user_data' rows could be removed, sounds good - I'll removed it then.

The result will be:
```
<boinc_user_id_verification>
 <master_url>http://209.250.242.113/boincserver/</master_url>
 <msg>1 oweijfwoeijfw</msg>
 <signature>C/acKvj9QyY1hL4BQX5Pu9oUDTtowLPbjyYQR5N/G8ooZVzfLTtVuh/CluWYvsfIpftHnvBFFz4UWCp7fike7K2wrM4NmfnQjH1wwnIZNb4CqAXS5awhCLxlo7PoPrVCMeppt+SGAIn2jQ54COiJxzSLdzEVuFlpLaOOJgQnrTepqw8ETzL8FwWry+SzN/CIBPiEjkOnOWZ3n1qWuZkNHrNb+z6lNGtw3yorB6ZRKY+CgDv5ylSHpNtBdv/XOE3fjBhJaRus14mKX8VtIKmrGd5SdcBVXZxZ98q5dpcOHqf7buk0xLW6a8qoEAI9T7slU18hQ40emhyHzhJ82d5ARg==</signature>
</boinc_user_id_verification>
```

Reply:
Code currently doesn't show the result of your discussion. Also see my comment about tags in `$message_data`.

Reply:
I've removed the fields discussed above. https://github.com/BOINC/boinc/pull/2134/files#diff-91d6b04155bcf59e677280a695167c2fR111